operator/v1/types_ingress.go

```diff
@@ -0,0 +1,119 @@
+package v1
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ClusterIngress describes a discrete, managed ingress component for the
+// cluster. This includes a router, and may also include a load balancer and
+// external wildcard DNS records for that load balancer.
+type ClusterIngress struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   ClusterIngressSpec   `json:"spec,omitempty"`
+	Status ClusterIngressStatus `json:"status,omitempty"`
+}
+
+// ClusterIngressSpec is the specification of the desired behavior of the
+// ClusterIngress.
+type ClusterIngressSpec struct {
+	// IngressDomain is a DNS name suffix serviced by the ClusterIngress. If
+	// specified, this value:
+	//
+	// * Is added automatically to the status of Routes to inform downstream systems.
+	// * May be used for a wildcard DNS alias to any provisioned load balancer.
+	// * May be used to generate a default wildcard certificate.
+	//
+	// If nil, the default value from the cluster ingress config is used.
+	//
+	// Once set, this field cannot be changed.
```

Comment:
Why is this `*string` instead of `string`?

Reply:
Don't say `nil`, this is godoc for JSON.  Say `if unset` or `if not specified`

Reply:
Describe the difference between null and empty.

Reply:
Also, explain the thing about if you don't set this for the next cluster ingress it won't work.

Reply:
This field essentially enables a bundle of features:

1. publishing the value to route status ("router canonical hostname")
2. enables DNS management (when using a supported platform)
3. used as the basis for generating default wildcard certs as needed

We could have decoupled all those things (DNS seems inevitable and I started [a proposal for it](https://docs.google.com/document/d/1VuBowd1jTunXskScs8VXi8u-sND6THWuWcl1Gta2MOs/edit?usp=sharing) already)...

Is it valuable to be able to turn all of that off at once? That's the only possible meaning of an empty value today, and I'm not sure it makes any sense.

Reply:
Spoke offline and decided this doesn't need to be a pointer after all. Will update docs.

Reply:
Took another stab at this, added an outstanding question about how to cross-reference fields